### PR TITLE
gimbal added

### PR DESCRIPTION
**Pull request by javilinos**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Issue(s) this addresses   | #244 |
| ROS2 version tested on | Humble|
| Aerial platform tested on |Ignition|

---

## Description of contribution in a few bullet points

* Added ability to attach sensor to gimbal from json 

## Description of documentation updates required from your changes

* How to add gimbal to simulation

---

## Future work that may be required in bullet points

* Create gimbal as a separate model

### CODE REVIEW

**Comment by javilinos** [comment]
So, Gimbal is now added as follows: 
```json
    "drones": [
        {
            "model_type": "quadrotor_base",
            "model_name": "M300",
            "xyz": [
                0.0,
                0.0,
                0.0
            ],
            "rpy": [
                0,
                0,
                0
            ],
            "flight_time": 60,
            "payload": [
                {
                    "model_name": "gimbal",
                    "model_type": "gimbal",
                    "payload": {
                        "model_name": "hd_camera",
                        "model_type": "hd_camera"
                    }
                }
            ]
        }
``` 
Right now as it states it is tested:
- Add as many gimbals you want in pararel, the published topic to control a gimbal is:  ``` /drone_ns/sensor/gimbal_cmd ``` and accepts a ``` geometry_msgs/vector3 ``` to control in roll, pitch and yaw. We have to discuss this as it should be our defined standard.
- You can of course add aditional sensors without gimbaling. 
- Multi drone (3) simulation with every drone containing a gimbal and an aditional sensor each. (6 sensors in total), 3 of them gimbaled 

What you cannot do right now:
- Gimbaled gimbals (I don't know why we would want this). To do this we have to make python payload functions calls recursive (right now there is one level of recursiveness). Also modify sdf so that every position set is relative to the one element that contains it. 

@pariaspe  I couldn't make a separate sdf model for gimbal as I still didn't figure out how to pass the parameters from the .jinja to that sdf model in order to build the structure (have tried many things but none worked). As gimbal is only composed of links and joints, it is included in the quadrotor.sdf.jinja. (Before you ask, ball or rev2 joints do not support joint position controllers). This is probably my main concern.

**Comment by pariaspe** [comment]
Hi @javilinos , without checking deeply all code changes, here are my first comments:
1. I personally don't like the approach of nesting payloads. I would include gimbals as a new attribute of `Payload` named `mount`. Default mount could be `fixed` and `mount: gimbal` will place your payload on a gimbal. Topic naming is also important, we should discuss a default naming, but I don't think we should allow the user to choose it on the JSON payload description. 
2. Gimbal command message should be Vector3Stamped, stamp and frame are important. Otherwise, vector is not clearly defined in space.
3. I agree, not dealing with gimbaled gimbals. If needed a new model should be created by the user.
4. There is an implementation of a [gimbal](https://github.com/gazebosim/gazebo-classic/blob/gazebo8/plugins/GimbalSmall2dPlugin.cc) in Gazebo classic, which is [not available yet](https://gazebosim.org/docs/fortress/comparison#model-plugins) in new Gazebo. However, their approach it is much interesting that ours. Take a look at [Model Plugins](https://classic.gazebosim.org/tutorials?tut=plugins_model&cat=write_plugin), at long term we should do something similar.
5. About jinja and sdf generation, it might change a bit after my previous comments. I don't like adding gimbal directly to the drone model. There is new [python bindings to create/edit sdf models](https://vimeo.com/879001733/c2df0a1823), which should allow what you were trying to do. 
7. How is the gimbal control done? PIDs? Or there isn't any control?
8. Why gimbal status is published in three different topics (Double msg)? Wouldn't it be better to report gimbal status with `vector3d` msgs?

**Comment by javilinos** [comment]
Hello @pariaspe, regarding your concerns:
1. By nesting Payloads the way we are doing it right now we can control the pose relative to base_link of the gimbal, as well as the pose relative to the gimbal of the attached sensor. Otherwise, the pose of the gimbal will always be the same. 
2. Understood, VectorStamped will be then. 
3. -
4. Building custom plugins could be a step forward in the future but right now we have all the necessary plugins to control the gimbal.
5. Me neither, It was the only thing I disliked about this approach, @RPS98 and me were looking into these python bindings to build sdf models vía python structures, but we have to check compatibilities with the Ignition libraries version that we are using, as well as the sdf format version that it generates that could be incompatible with out simulator version. I think this could be a good long term update into more intuitive and readable models. Will look into it. 
6. The control is done by publishing a ```geometry_msgs/Vector3``` message into the /drone_ns/sensor_model/gimbal_cmd. Each value of the vector perform position control in r, p, y respectively. 
7. Ball joints are a type of joint that can perform rotation in these three axis, however, position control plugin does not support it. The only way of getting the gimbal done was creating a structure like: base_link->roll_joint -> pitch_link->pitch_joint->yaw_link->yaw_joint->sensor. As it has three separate joints, there is a (PID) position controller for each joint, thats why you are seeing the status for each joint. The custom bridge takes the Vector3 message and republish each value into the respective controller. We should disable the publishing of the states done by jointStatePublisher plugin, and move that to the bridge to publish it as a vector3 message as you are saying.

**Comment by RPS98** [comment]
Hi @pariaspe and @javilinos, some comments:

1. I like the idea of allowing the user to define the gimbal's pose relative to the aircraft body and the sensor's pose relative to the gimbal. These parameters are easily measurable. I believe we should indeed define a translation between the gimbal's pivot point and the sensor's position because gimbal rotations will also cause translations of the sensor relative to the aircraft body, which is realistic. If we enforce the sensor and gimbal to be at the same point, we are limiting the simulation options significantly.
2. Regarding the message type to use for control, we should look into [PX4 Gimbal Control Msg](https://github.com/PX4/px4_msgs/blob/main/msg/GimbalControls.msg), [PX4 Gimbal Attitude Msg](https://github.com/PX4/px4_msgs/blob/main/msg/GimbalDeviceSetAttitude.msg), [PX4 Gimbal Attitude Manager Msg](https://github.com/PX4/px4_msgs/blob/main/msg/GimbalManagerSetAttitude.msg), and [DJI PSDK Gimbal Msg](https://github.com/umdlife/psdk_ros2/blob/main/psdk_interfaces/msg/GimbalRotation.msg). All of them include a header indicating the `timestamp` and `frame_id` (gimbal to be rotated), an integer for control mode, and a vector3 or 3 independent floats indicating which axes the control applies to. Control modes include position (relative or absolute) and angular velocity.
7. Concerning the message type to use in aerostack2 for status, there are [PX4 Gimbal Status Msg](https://github.com/PX4/px4_msgs/blob/main/msg/GimbalDeviceAttitudeStatus.msg) and [DJI PSDK Gimbal Status Msg](https://github.com/umdlife/psdk_ros2/blob/main/psdk_interfaces/msg/GimbalStatus.msg). I believe it's best to leave fields for absolute position and angular velocity, and depending on the gimbal, the user fills in one or the other fields. With this, we'll need to define what type of message the gazebo platform should output.

**Comment by javilinos** [comment]
@RPS98 @pariaspe there is also a couple of things we have to agree on:
1. Topic naming. Right now, it is:  ``` /drone_ns/sensor_name/gimbal_cmd/orientation ``` for position control and  ``` /drone_ns/sensor_name/gimbal_cmd/speed ``` for speed control. 
2. It is necessary to define a custom message other than using vector3Stamped? @RPS98 suggested adding some fields into a gimbal message so if a drone contains multiple gimbals user could select which gimbal to control, in that case, topic naming could be just  ``` /drone_ns/gimbal/orientation ``` or ``` /drone_ns/gimbal/twist ```
3. Control mode is chosen depending on what gimbal type defined in json. Right now Ignition does NOT support two types of joint controllers (position and speed) simultaneously in a single joint. 
4. Gimbal status is not published yet, but should be something like ``` /drone_ns/sensor_measurements/gimbal_x ```

What @RPS98 is saying makes sense, if there is a sensor is translated from the gimbal's 0,0, then when the gimbal rotates this results in a rotation but also a translation of the sensor.

**Comment by pariaspe** [comment]
@RPS98 
1. I don't fully understand it, but if you do it's ok for me :+1: 
2. Agreed
3. Do we need angular speed too? Publishing pose at a certain hz wouldn't be enough?

@javilinos 
1. Let's try to define them. Suggestions are welcome:

| Topic name | Msg Type | Description |
|---|---|---|
| `/drone_ns/sensor_name/gimbal_command/pose/orientation` | (a) [Vector3Stamped](http://docs.ros.org/en/melodic/api/geometry_msgs/html/msg/Vector3Stamped.html) or (b) [QuaternionStamped](http://docs.ros.org/en/melodic/api/geometry_msgs/html/msg/QuaternionStamped.html) | Topic to command a gimbal via position control |
| `/drone_ns/sensor_name/gimbal_command/twist/angular` | [Vector3Stamped](http://docs.ros.org/en/melodic/api/geometry_msgs/html/msg/Vector3Stamped.html) | Topic to command a gimbal via speed control |
| `/drone_ns/sensor_measurements/gimbal_name/pose/orientation`  | [QuaternionStamped](http://docs.ros.org/en/melodic/api/geometry_msgs/html/msg/QuaternionStamped.html) | Angular position status topic  |
| `/drone_ns/sensor_measurements/gimbal_name/twist`  | [TwistStamped](http://docs.ros.org/en/melodic/api/geometry_msgs/html/msg/TwistStamped.html) | Angular speed status topic |

2. See above table. It might be easier to define a custom message and collapse topics, obtaining one status and one commander topic.
3. Alright, I am wondering why you divided gimbal into two different enum.
4. Again see above table.

**Comment by pariaspe** [comment]
> @pariaspe about your question, position and speed controllers working at the same time does not work, so we need to differentiate from the json what type of gimbal we are using. This could be done vía defining two different names for gimbal such as the ones defined, or add a gimbal_type in case a gimbal is defined.
> 
> Naming seems logic to me, however in case a drone contains two or more gimbals, where do we specify what gimbal is the control command given to, the frame_id of vector3Stamped? That's why @RPS98 proposed to create a gimbalControl message

Two different enum values but one Gimbal class, there is no need to have two classes.

I didn't include gimbal name at the control topics since it's linked to one sensor, but it can also be added. A new msg also looks good to me :+1:

**Comment by RPS98** [comment]
There are two distinct components: the Aerostack2 gimbal standard and the Gazebo bridge.

For the Aerostack2 gimbal control and status standards, I propose the following:

| Topic name | Msg Type | Description |
|---|---|---|
| `/drone_ns/platform/gimbal_command` | Custom MSG | Message including a timestamp, gimbal ID, control mode, and axis values. |
| `/drone_ns/sensor_measurements/gimbal_name/attitude`  | [QuaternionStamped](http://docs.ros.org/en/humble/api/geometry_msgs/html/msg/QuaternionStamped.html) | Angular position status topic  |
| `/drone_ns/sensor_measurements/gimbal_name/twist`  | [Vector3Stamped](http://docs.ros.org/en/humble/api/geometry_msgs/html/msg/Vector3Stamped.html) | Angular speed status topic |

This approach consolidates gimbal control under a single topic, incorporating the gimbal ID within it. I believe this consolidation would facilitate the use of a unified callback on the aerial platform through a virtual method.

Regarding retrieving gimbal status, two topics are proposed. Depending on the scenario, one of these topics may not be published.

**Comment by javilinos** [comment]
@RPS98 @pariaspe for now the state of the gimbal development is as follows:
- Gimbal working in position and speed mode in a separate way. It cannot work in both modes simultaneously. 
- Despite the previous point, there is a control mode selector in the as2 custom message. If the gimbal does not support the selected control mode, the platform will Just give you a warning and do nothing.
- Speed and attitude states published as: ``` /drone0/sensor_measurements/{gimbal_id}/twist ``` and ```  /drone0/sensor_measurements/{gimbal_id}/attitude ``` of type [QuaternionStamped](http://docs.ros.org/en/humble/api/geometry_msgs/html/msg/QuaternionStamped.html) and [Vector3Stamped](http://docs.ros.org/en/humble/api/geometry_msgs/html/msg/Vector3Stamped.html) respectively
- Custom msg GimbalControl.msg added, composed of 
  ``` 
  uint8 SPEED_MODE = 0
  uint8 POSITION_MODE = 1

  geometry_msgs/Vector3Stamped control # x: roll y: pitch z: yaw
  uint8 control_mode ```
- Topic to control is ``` /{ns}/platform/{gimbal_id}/gimbal_command ``` 
- frame_id is still not set in status messages. Position is set respect to it's initial position (so initial position would be 0,0,0), speed would be respect to itself. 

We thought about not including it in aerial platform yet. 

In order to be included in as2_core/names, states topics should include a differentiable static part. such as ``` /drone0/sensor_measurements/gimbal/{gimbal_id}/twist ``` and ```  /drone0/sensor_measurements/gimbal/{gimbal_id}/attitude ``` 

Only thing I dislike about this implementation is the joints from which the gimbal is built are included in the quadrotor.sdf.jinja. It would be easy to take it out from it with python bindings, but without it I still cannot think about a way to get it out. I will revisit it.

A new simple behavior is going to be built to control the gimbal, maybe it could be called "move_gimbal" with some additional functions to move it in position mode and speed mode.

**Comment by miferco97** [comment]
> * Speed and attitude states published as: `/drone0/sensor_measurements/{gimbal_id}/twist` and ` /drone0/sensor_measurements/{gimbal_id}/attitude` of type [QuaternionStamped](http://docs.ros.org/en/humble/api/geometry_msgs/html/msg/QuaternionStamped.html) and [Vector3Stamped](http://docs.ros.org/en/humble/api/geometry_msgs/html/msg/Vector3Stamped.html) respectively

I would prefer to have the whole state with one single stamp, so I vote for creating a GimbalStatus msg with a Header, the speed and the orientation

**Comment by pariaspe** [comment]
Nice work @javilinos !

> Only thing I dislike about this implementation is the joints from which the gimbal is built are included in the quadrotor.sdf.jinja. It would be easy to take it out from it with python bindings, but without it I still cannot think about a way to get it out. I will revisit it.

About this, I'll take a look at it too.

File: as2_simulation_assets/as2_ign_gazebo_assets/src/ign_assets/models/drone.py [hunk 1]
```diff
@@ -201,10 +215,16 @@ def generate(self, world) -> tuple[str, str]:
...
-        for line in err_output.splitlines():
-            if line.find('undefined local') > 0:
-                raise RuntimeError(line)
+        stdout, stderr = process.communicate()
+
+        # You can now access the captured output
+        print("Standard Output:")
+        print(stdout.decode('utf-8'))  # Convert bytes to string
+            
+        # stderr = process.communicate()[1]
```
Comment (pariaspe) [inline as2_simulation_assets/as2_ign_gazebo_assets/src/ign_assets/models/drone.py:224]: restore exception management

File: as2_simulation_assets/as2_ign_gazebo_assets/models/quadrotor_base/quadrotor_base.sdf.jinja [hunk 2]
```diff
@@ -177,37 +176,156 @@
             <power_load>6.6</power_load>
             <start_on_motion>true</start_on_motion>
             <fix_issue_225>true</fix_issue_225>
+        </plugin> {% endif -%} {% for sensor in
```
Comment (pariaspe) [inline as2_simulation_assets/as2_ign_gazebo_assets/models/quadrotor_base/quadrotor_base.sdf.jinja:179]: missing for hexrotor

File: as2_simulation_assets/as2_ign_gazebo_assets/src/ign_assets/world.py [hunk 3]
```diff
@@ -206,40 +206,26 @@ def dummy_world() -> World:
...
+                    "model_type": "hd_camera"
                 }
             ]
         }
         ]
     }
     """
     world_model = World.parse_raw(WORLD_JSON)
-
+    print(world_model.drones)
```
Comment (pariaspe) [inline as2_simulation_assets/as2_ign_gazebo_assets/src/ign_assets/world.py:227]: do some testing instead of logging

**Review by pariaspe — CHANGES_REQUESTED** [review]

**Review by pariaspe — APPROVED** [review]

**Review by pariaspe — APPROVED** [review]

**Pull request merged**